design/baremetal-operator/bmh_non-bootable_iso.md

```diff
@@ -0,0 +1,124 @@
+to a system after provisioning has finished, for example :
+
+A configuration ISO can be used to provide first-boot configuration to a host
+deployed from a standard image in advance. An operator may have a pool of
+powered off hosts with operating system already provisioned. Then, adding
+a host to the cluster boils down to powering it on with a configuration ISO provided.
+
+### Goals
+
+Expose the Ironic API to attach non-bootable iso images via Metal3.
+
+### Non-Goals
+
+Supporting non-ISO images i.e. media types other than CD/DVD, supporting
+drivers other than Redfish and its derived drivers.
+
+## Proposal
+
+### BMO Proposal
+
+Add a new optional section in the BareMetalHost spec containing details of the
+non-bootable iso image, example :
+
+#### Design spec
+
+```yaml
+  spec:
+    dataImage:
+      url: http://1.2.3.4/image.iso
+```
+
+We may consider adding support for HTTP credentials and TLS settings in the future
+but it's outside the scope of this proposal.
+
+**Note** : As part of this proposal, the dataImage will be attached after the node
+has been Provisioned or ExternallyProvisioned.
+
+### Ironic Support
+
+The [proposal](https://bugs.launchpad.net/ironic/+bug/2033288) for exposing this functionaility in Ironic has been
+accepted by the upstream Ironic community.
+
+The [Ironic implementation](https://review.opendev.org/c/openstack/ironic/+/894918) based on the above proposalon has been completed.
+
+Other than that, we just need the implementation for driver specific support,
+which in this proposal is Redfish.
+
+## Design Details
+
+Add an optional section `DataImage` to `BareMetalHostSpec`, with sub-field
+`Url`.
+
+The ISO will be attached when the `BareMetalHost` object is edited to add
+the `DataImage` spec triggering a reconciliation. In case a reboot is requested
+at the same time using `RebootAnnotation`, the ISO will be attached first.
+
+The `BareMetalHostStatus` will cache the image details, which will inform
+us if the attachment succeeded and will also be used for detachment, either
+when the attachment fails and we retry or when the user explicitly requests
+detachment by removing the `DataImage` spec from `BareMetalHost` object.
+
+A flag will be used at the webhook level to validate if a given driver supports the
+non-bootable ISO feature. Since the support for attaching an ISO as virtual media is
+equivalent for both `PreprovisioningISOImage` and `NonBootableISOImage`, we plan to
+rename the existing flag for `PreprovisioningISOImage` i.e. `SupportsISOPreprovisioningImage`
+to `SupportVirtualMediaISOImage` and use it for both the cases.
+
+### Implementation Details
+
+As part of the design we are assuming that the user will edit the BareMetalHost
+to add the dataImage spec, and might also request reboot via rebootAnnotation
+or power state change via `online` field.
+
+We want to attach the dataImage first, so handling the attachment/detachment of dataImage
```

Comment:
Should we block the attachment of a DataImage for provisioned BMHs explicitly?
I think yes, or at least have an annotation or flag or some option to disable this for provisioned BMHs, because this seems to a me a potential source of issues.

Reply:
Why so? It's actually one of the potential use cases of this API (although not literally the one we're looking at).

Reply:
Well, until now we didn't really had any functionality that would modify the content of a provisioned BMH, we are able to detach it and re-provision/clean it but no functionality that would modify the block devices/data present on a machine without re-provisioning. 

In many scenarios I agree that on the fly media attachment could be a useful feature but then there could be other scenarios where users would like to make sure that  no block device gets attached to their node not even by accident after it was provisioned. 

A very simple example is why you could disable USB and other I/O ports on a computer and  put a password on the BIOS, there are operating environments where the possibility of attaching any additional I/O device is viewed as a potential attack vector. 

I agree with you that we should allow attaching non bootable iso to the already provisioned machines, but I would really like an option to disable this for people who see this as a potential attack vector.

Reply:
> Well, until now we didn't really had any functionality that would modify the content of a provisioned BMH

There is definitely interest in changing BIOS  settings and doing firmware upgrades for deployed hosts, btw. Ironic is getting there in terms of features.

> the possibility of attaching any additional I/O device is viewed as a potential attack vector.

If someone has access to your BMH, your security is already breached.

Reply:
Yes but I don't only mean this for security reasons, for example there could be a user error and someone accidentally updates the BMH resource (e.g. right now BMO sees all BMH resources in all namespaces ) , my main reason would be to have the option producing "finalized" BMHs.

Reply:
It's a valid concern, but how is it specific to this feature? You can accidentally deprovision a BMH by nuking its image field.

Doesn't detached state solve the issue?

Reply:
You are totally right that user error could cause deprovisioning, or re-inspection or something, the issue I see here is that switching the block device (blk from now on) to something else or injecting a blk on the machine rom an other user (for example in multy cluster/user environments) could cause the following issues: sharing the data of an other custommer/user, damaging data integrity in general without noticing ( an accidentally attached block device could modify the behavior of the business logic of the end user even).

Until now  our errors  either caused the machine to not upgrade, not provision or the user might have accidentally deprovisioned or "broke" the machine but with this feature we open up a whole host of data integrity related questions.

Detached annotation only partially solves this problem as that would remove the BMH from management which causes loss of functionality (sure you can bring it back but that would add extra steps seems to be a bit heavy handed)  but I am talking about use cases where user would like to manage the machines just as usual, the only thing the user would want is to avoid any threats to data integrity, a machine being accidentally deprovisioned or broken are  "self contained" issues from the perspective of the data on the machine, the data might be lost, but there is no threat of getting someone else's data, or getting data that would effect the workloads and they would do something unwanted, with this feature these scenarios would all become possible.

From my POV an annotation would be fine to block attaching blk and block modifying the attached blk. Ofc I am also open to other ways that we could provide data integrity protection.

TLDR, attaching an incorrect blk to a running BMH might cause way bigger damage than losing that specific machine.

Reply:
> Until now our errors either caused the machine to not upgrade, not provision or the user might have accidentally deprovisioned or "broke" the machine but with this feature we open up a whole host of data integrity related questions.

If you have write access to the BMH you can disable cleaning and reprovision with your own image.
For that matter, you could remove this new annotation you're proposing.

I don't think there is any substitute for using RBAC to not give write access to the BMH to users who aren't fully trusted.

Reply:
Yeah. Ironic itself has RBAC that could be used to e.g. allow some users to provision but not attach ISO or change BIOS settings. I guess we don't have anything like that?

Would it help if we introduced a new CR for attaching ISOs? Then an operator could disable everyone's access to such resources?

Reply:
This CR + RBAC based solution sounds the best so far, I am fine with it ! Thanks!

Reply:
Made changes to design to use CR + Rbac for dataImage